### DISCUSSION TITLE
Compile time sum types

### DISCUSSION THREAD

**Discussion by mcastorina** [vlang/v#11894] [Ideas]
Currently sum types are checked at run-time, but the code can be generated in a way that does not require these checks. The trade off would be more generated code / possibly longer compilation time, so it should be a complementary feature.

Simple example:

```v
type Sum = int | string

fn test(x Sum) int {
    match x {
        int { return 0 }
        string { return 1 }
    }
}

test(123)
test('hello')
```

would be logically equivalent to

```v
fn test_int(x int) int {
    return 0
}
fn test_string(x string) int {
    return 1
}

test_int(123)
test_string('hello')
```

**Comment by crthpl** [comment]
This is probably what generics will be when there are constraints (You should be able to use both sum types and interfaces as constraints)
And they already work like this if you just do:
```
fn test<T>(x T) int {
  $if T is int {
    return 0
  } $else $if T is string {
    return 1
  }
```
Compile-time matches still need implementing though.

**Comment by mcastorina** [comment]
Yeah, perhaps a sum-type can be thought of as a form of constraints on generics.

**Comment by danieldaeschle** [comment]
sum types are tagged unions. There is no need to change this :)
I hope you're okay if I close the issue.

**Comment by mcastorina** [comment]
I'm not suggesting sum types change, but a feature to be added that will separate the branches for possible sum-type variants at compile time. This makes the run-time cost of sum-types zero.

As @crthpl said, this is effectively generics with constraints.

I think this issue should be kept open until it is possible to constrain a generic with a sum-type.

**Comment by danieldaeschle** [comment]
This issue is about generic constraints: https://github.com/vlang/v/issues/2711

**Comment by dumblob** [comment]
Wait, that's two different things - "generic constraints" are by far not the same as "sum types + conditional compilation". Please don't confuse these two. Yes, with "sum types + conditional compilation" one can achieve semantically something very similar (even the same in some cases) as "generic constraints". But this absolutely doesn't work the other way round (and this is also one of the reasons I'm arguing against generic constraints in the linked thread).

I for myself am against "generic constraints" but I'm all for "sum types (incl. compile-time sum types) + conditional compilation".

**Comment by danieldaeschle** [comment]
How would @mcastorina's example work with this?:

```v
type Sum = int | string

fn test(x Sum) int {
    match x {
        int { return 0 }
        string { return 1 }
    }
}

test(Sum('hello'))
```

**Comment by mcastorina** [comment]
My idea is that the generated C code would have one function for each variant of the sum-type, and the call site uses the appropriate version.

```c
int test_string(string x) {
    return 1;
}
int test_int(int x) {
    return 0;
}

test_string(_SLIT("hello"));
```
Perhaps there are other ways to achieve this. Maybe it would be better to transform the AST directly so it's not limited to the C backend.

**Comment by danieldaeschle** [comment]
You didn't answer my question. Which function would it use when the type is not known at compile time?

**Comment by mcastorina** [comment]
The type is always known at compile time. V is a typed language.

**Comment by JalonSolov** [comment]
Yes, please... transform the AST so all backends can do the same thing.  A lot of stuff in cgen already needs to be moved to the AST for the same reason.

**Comment by danieldaeschle** [comment]
>The type is always known at compile time. V is a typed language.

That's wrong.

```v
type Alphabet = Abc | Xyz

mut a := Alphabet(Abc{}) // the compiler could know that a is Abc but it's castet to Alphabet so the compiler knows it's of type Alphabet

go my_thread_func(mut a)

test(a) // how should the compiler know the type variant here? it could have changed in another thread
```

**Comment by crthpl** [comment]
@danieldaeschle The code generation for the `test(a)` could be:
```c
if (a._typ == 7 /* int type idx */) {
    test_int(a->int)
} else if (a._typ == 18 /* string type idx */) {
    test_string(a->string)
}
```
or something like that.

**Comment by mcastorina** [comment]
@danieldaeschle that's a good point, and in your example the sum-type is acting as a tagged union (as it is currently). This means the sub-type is handled at run-time, but the type *is* `Alphabet`.

I don't think compile-time sum-types would be able to handle your example. The main problem I see is a variable being able to change it's sub-type at run-time. Like I said though, this is not aiming to replace the current "tagged union" sum-types.
